File: algo/108. Convert Sorted Array to Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if len(nums) == 0:
            return
        if len(nums) == 1:
            return TreeNode(nums[0])
        if len(nums) == 2:
            node1 = TreeNode(nums[0])
            node2 = TreeNode(nums[1])
            node2.left = node1
            return node2
        if len(nums) == 3:
            node1 = TreeNode(nums[0])
            node2 = TreeNode(nums[1])
            node3 = TreeNode(nums[2])
            node2.left = node1
            node2.right = node3
            return node2
        midIndex = len(nums) // 2
        mid = TreeNode(nums[midIndex])
        left = self.sortedArrayToBST(nums[:midIndex])
        right = self.sortedArrayToBST(nums[midIndex+1:])
        mid.left = left
        mid.right = right
        return mid

File: round3/108. Convert Sorted Array to Binary Search Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode SortedArrayToBST(int[] nums) {
        return Helper(nums, 0, nums.Length - 1);
    }
    
    private TreeNode Helper(int[] nums, int start, int end) {
        if (end < start) return null;
        
        int mid = (end - start) / 2 + start;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = Helper(nums, start, mid - 1);
        root.right = Helper(nums, mid + 1, end);
        return root;
    }
}